docs(adr): revise ADR-20: Post ownership transfer (#1121)
desmos-labs/desmos: docs/architecture/adr-020-post-ownership-change.md

## Changelog
- April 21th, 2023: First draft;

> - April 26th, 2023: First revise;

Reviewer: ```suggestion
- April 26th, 2023: First review;
```

## Decision

> We will add an __owner__ field to the `Post` structure and implement post ownership transfer functionality. Simple transfer functionality could result in someone receiving spam messages that damage their reputation. For example, scammers can transfer scam messages to victims, leading to reputational harm. Therefore, the functionality we will implement should include the following features:
> 1. Permit the sender to initiate a request to transfer post ownership to the receiver
> 2. Allow the receiver to accept or refuse the transfer request from the sender
> 3. Allow the sender to cancel the transfer request if the receiver has not deal with the request

Reviewer: I've made it more readable using ChatGPT
```suggestion
We will enhance the `Post` structure by adding an **owner** field and implementing post ownership transfer functionality. It's important to consider potential issues that could arise from simple transfer functionality, such as someone receiving spam messages that could harm their reputation. For instance, scammers may transfer scam messages to victims, leading to reputational damage. To address this, the implemented functionality will include the following features:

1. **Initiating Transfer**: The sender will be able to initiate a request to transfer post ownership to the receiver.
2. **Accepting or Refusing Transfer**: The receiver will have the option to accept or refuse the transfer request from the sender.
3. **Canceling Transfer Request**: In case the receiver has not yet addressed the transfer request, the sender will be able to cancel it.

```

  // Owner of the post
  string owner = 14;
}

In order to handle the post owner transfer process easily, we will define a new structure as follows:

> // PostOwnerTransferRequest represent a request that transfers the post ownership to receiver

Reviewer: ```suggestion
// PostOwnerTransferRequest represents a request to transfer the ownership of a post from the sender to the receiver
```

message PostOwnerTransferRequest {

>   // Id of the subspace where the post owner transfer will be performed

Reviewer: ```suggestion
  // Id of the subspace that holds the post to transfer
```

  1. request a post owner transfer to a receiver
  2. cancel a post owner transfer request
  3. accept a post owner transfer request
  4. refuse a post owner transfer request

```proto
service Msg {
  // RequestPostOwnerTransfer allows sender to send a request to transfer a post ownership to receiver
  rpc RequestPostOwnerTransfer(MsgRequestPostOwnerTransfer) returns (MsgRequestPostOwnerTransferResponse);

  // CancelPostOwnerTransfer allows sender to cancel an outgoing post owner transfer request
  rpc CancelPostOwnerTransfer(MsgCancelPostOwnerTransfer) returns (MsgCancelPostOwnerTransferResponse);

  // AcceptPostOwnerTransfer allows receiver to accept an incoming post transfer request
  rpc AcceptPostOwnerTransfer(MsgAcceptPostOwnerTransfer) returns (MsgAcceptPostOwnerTransferResponse);

  // RefusePostOwnerTransfer allows receiver to refuse an incoming post transfer request
  rpc RefusePostOwnerTransfer(MsgRefusePostOwnerTransfer) returns (MsgRefusePostOwnerTransferResponse);
}

// MsgRequestPostOwnerTransfer represent a message used to transfer a post ownership to receiver
message MsgRequestPostOwnerTransfer {

>   // Id of the subspace where the post owner transfer will be performed

Reviewer: ```suggestion
  // Id of the subspace that holds the post which ownership should be transfered
```

  uint64 subspace_id = 1;
    
  // Id of the post which will be transferred
  uint64 post_id = 2;

  // Address of the post ownership receiver
  string receiver = 3;
    
  // Address of the sender who is creating a transfer request
  string sender = 4;
}
// MsgRequestPostOwnerTransferResponse defines the Msg/RequestPostOwnerTransfer response type
message MsgRequestPostOwnerTransferResponse {}

// MsgCancelPostOwnerTransfer represents a message used to cancel a outgoing post transfer request
message MsgCancelPostOwnerTransfer {

>   // Id of the subspace where the request will be cancelled

Reviewer: ```suggestion
  // Id of the subspace that holds the post for which the request should be canceled
```

  uint64 subspace_id = 1;
    

>   // Id of the post whose request will be cancelled

Reviewer: ```suggestion
  // Id of the post for which the request will be cancelled
```

  uint64 post_id = 2;

>   // Address of the sender who is cancelling the request

Reviewer: ```suggestion
  // Address of the transfer request sender
```

  string sender = 3;
}
// MsgCancelPostOwnerTransferResponse defines the Msg/CancelPostOwnerTransfer response type
message MsgRequestPostOwnerTransferResponse {}

// MsgAcceptPostOwnerTransfer represents a message used to accept a incoming post transfer request
message MsgAcceptPostOwnerTransfer {

>   // Id of the subspace where the request will be accepted

Reviewer: ```suggestion
  // Id of the subspace holding the post for which the request will be accepted
```

  uint64 subspace_id = 1;
    

>   // Id of the post whose request will be accepted

Reviewer: ```suggestion
  // Id of the post for which the request will be accepted
```

  uint64 post_id = 2;

>   // Address of the receiver who is accepting the request

Reviewer: ```suggestion
  // Address of the request receiver
```

  string receiver = 3;
}

// MsgAcceptPostOwnerTransferResponse defines the Msg/AcceptPostOwnerTransfer response type
message MsgAcceptPostOwnerTransferResponse {}

// MsgRefusePostOwnerTransfer represents a message used to refuse a incoming post transfer request
message MsgRefusePostOwnerTransfer {

>   // Id of the subspace where the request will be refused

Reviewer: ```suggestion
  // Id of the subspace holding the post for which the request will be refused
```

  uint64 subspace_id = 1;
    

>   // Id of the post whose request will be refused

Reviewer: ```suggestion
  // Id of the post for which the request will be refused
```

  uint64 post_id = 2;

>   // Address of the receiver who is refusing the request

Reviewer: ```suggestion
  // Address of the request receiver
```

### Positive

> - Enable users to transfer a post to another user

Reviewer: ```suggestion
- Enable users to transfer the ownership of a post to another user
```
